File: frontend/src/types/risk.ts
/** Risk-engine types - mirror backend/app/schemas/risk.py exactly. */

export type RiskLevel = "LOW" | "MODERATE" | "HIGH" | "CRITICAL";

export interface RiskZone {
  zone_id: number;
  center_latitude: number;
  center_longitude: number;
  radius_meters: number;
  point_count: number;
  incident_count: number;
  sos_count: number;
  dominant_incident_type: string;
  avg_severity: number;
  distinct_types: number;
  last_event_at: string;
  risk_score: number;
  risk_level: RiskLevel;
}

export interface RiskZonesResponse {
  generated_at: string;
  eps_km: number;
  min_samples: number;
  total_incidents: number;
  total_sos: number;
  noise_incidents: number;
  noise_sos: number;
  zone_count: number;
  zones: RiskZone[];
}

export interface HeatmapMarker {
  zone_id: number;
  /** [latitude, longitude] - Leaflet order. */
  center: [number, number];
  radius_meters: number;
  risk_score: number;
  risk_level: RiskLevel;
  point_count: number;
  incident_count: number;
  sos_count: number;
  dominant_incident_type: string;
  avg_severity: number;
  last_event_at: string;
}

export interface RiskHeatmapResponse {
  generated_at: string;
  eps_km: number;
  min_samples: number;
  total_incidents: number;
  total_sos: number;
  noise_incidents: number;
  noise_sos: number;
  marker_count: number;
  markers: HeatmapMarker[];
}

File: backend/app/schemas/risk.py
"""Pydantic schemas for the DBSCAN-based geographic risk engine."""
import datetime
from typing import Literal

from pydantic import BaseModel, Field

RiskLevel = Literal["LOW", "MODERATE", "HIGH", "CRITICAL"]


class RiskZone(BaseModel):
    """A single geographic risk zone produced by DBSCAN clustering."""

    zone_id: int = Field(..., ge=1, description="1-based rank, ordered by risk_score desc")
    center_latitude: float = Field(..., ge=-90, le=90)
    center_longitude: float = Field(..., ge=-180, le=180)
    radius_meters: float = Field(..., ge=0, description="Max distance from center to a member")
    point_count: int = Field(..., ge=1)
    incident_count: int = Field(..., ge=1)
    sos_count: int = Field(..., ge=0)
    dominant_incident_type: str = Field(..., min_length=1)
    avg_severity: float = Field(..., gt=0, le=5, description="Mean raw severity of members")
    distinct_types: int = Field(..., ge=1)
    last_event_at: datetime.datetime
    risk_score: int = Field(..., ge=0, le=100)
    risk_level: RiskLevel


class RiskZonesResponse(BaseModel):
    """Envelope returned by GET /api/risk-zones."""

    generated_at: datetime.datetime
    eps_km: float = Field(..., gt=0)
    min_samples: int = Field(..., ge=1)
    total_incidents: int = Field(..., ge=0)
    total_sos: int = Field(..., ge=0)
    noise_incidents: int = Field(..., ge=0, description="Incidents not assigned to any zone")
    noise_sos: int = Field(..., ge=0, description="SOS events not assigned to any zone")
    zone_count: int = Field(..., ge=0)
    zones: list[RiskZone]